### PR TITLE
[Very WIP] [Probably horrifically broken] "Smart" high-level music player

### PR DESCRIPTION
**Pull request by benjamin-allen**
The idea is to create a series of methods that evaluate certain aspects of the game and determine an appropriate song to play at the given time. However, accessing this feature should be a high-level system, preferably something to be done in .json files.

What's desired:
- [x] Get and play music based on the player's location
  - [x] Play special music for certain overmap tiles. E.g, have music specifically for labs
  - [x] Make these overmap-specific playlists easy to define for the end user. Preferably from musicset.json
- [x] If there's no special music playing, combine various factors to select various songs
  - [x] Possibly combine danger level, current weather, date/time, etc.
  - [ ] This would probably necessitate modifying how playlists are read to include more parameters to match these songs
- [x] If no song can be found for the ambient conditions, fallback to some kind of default
  - [x] One possibility: Play songs based on the player's proximity to a city (Currently, this is what's being done instead of the factor combination)
  - [x] Another: Randomly select an ambient song
- [x] If absolutely nothing else can be found, play from a fallback playlist

What's done:
- Music can be played from a `wilderness`, `outskirts`, or `city` playlist
- A function is implemented to play from a `fallback` playlist, but it is unused
- There is a method that plays a song and also returns an integer value as to the status of the song (-1 for "no such playlist", 0 for "song is already playing", 1 for "playing song now") This will probably be changed to a bool.

Problems:
- All of the current work is done during game::do_turn, which may be more often than necessary.
- The boolean variant of play_music required the entire function to be copypasted with some different returns.

Tested:
- Tested on 64-bit Linux with make TILES=1 SOUND=1 CACHE=1 LOCALIZE=0
- NOT tested on a build where TILES=0 yet

### CODE REVIEW

**Comment by kevingranade** [comment]
If you're installing hooks in a lot of places like it looks like you are
going to, it'd be really nice to add them as generic hook methods and put
your music changing logic there instead of inline in the regular code.
Ideally this kind of thing would be supported by an observer system where
code can register itself to receive notifications about game events.

**Comment by Nullsrc** [comment]
I would have to learn how to do it, but I can try looking into it. I'll definitely end up asking lots of questions about hooks and whatnot, though. 😆

**Comment by DanmakuDan** [comment]
Related to the astyle regression: look in `doc/CODE_STYLE.txt`, it has the astyle command to run to format your file.

**Comment by lucianposton** [comment]
Related issue to keep in mind with this PR: https://github.com/CleverRaven/Cataclysm-DDA/issues/16258

**Comment by Nullsrc** [comment]
This is about as done as I can make it, and I definitely won't have any more time to work on it. Merge if it's up to par, close otherwise.

File: src/sounds.cpp [hunk 1]
```diff
@@ -952,6 +953,75 @@ void sfx::do_obstacle() {
...
+            return play_special_music( "d_horde" );
+        } 
+    }
+    prev_hostiles = hostiles;
+    return -1;
+}
+
+void sfx::play_city_distance_music() {
+    const auto wild_fctr = overmap_buffer.closest_city( g->u.global_sm_location() );
+    const auto &nearest_city = *wild_fctr.city;
```
Comment (BevapDin) [inline src/sounds.cpp]: `wild_fctr.city` needs to be checked for `nullptr`, it's possible that there is no nearest city at all.

**Comment by mugling** [comment]
> I definitely won't have any more time to work on it. Merge if it's up to par, close otherwise.

Stalled and currently unmergeable. Closed as per OP's comments

**Pull request closed without merging**